fabric-rendering-v1/src/client/java/net/fabricmc/fabric/impl/client/rendering/hud/HudStatusBarHeightRegistryImpl.java

```diff
@@ -0,0 +1,303 @@
+		LivingEntity livingEntity = gui.getRiddenEntity();
+		return gui.getHeartCount(livingEntity) == 0 ? 10 : 0;
+	};
+	/**
+	 * Height provider for the vanilla air bar.
+	 *
+	 * <p>Mods that would otherwise have a mixin for this are encouraged to instead register a replacement provider via
+	 * {@link HudStatusBarHeightRegistry#addRight(Identifier, ToIntFunction)}.
+	 */
+	static final ToIntFunction<PlayerEntity> AIR_BAR = (PlayerEntity player) -> {
+		int maxAirSupply = player.getMaxAir();
+		int airSupply = Math.clamp(player.getAir(), 0, maxAirSupply);
+		boolean isInWater = player.isSubmergedIn(FluidTags.WATER);
+		return isInWater || airSupply < maxAirSupply ? 10 : 0;
+	};
+
+	/**
+	 * Height providers used during rendering computed from everything that was registered.
+	 *
+	 * <p>These providers do NOT
+	 * return the heights of individual elements; instead they return the height at which an element should render at,
+	 * which is computed by summing all the heights from providers considered "below" an element.
+	 */
+	@Nullable
+	static Map<Identifier, ToIntFunction<PlayerEntity>> resolvedHeightProviders;
+
+	static {
+		ImmutableMap.Builder<Identifier, ToIntFunction<PlayerEntity>> builder = ImmutableMap.builder();
+		builder.put(VanillaHudElements.HEALTH_BAR, ZERO);
+		builder.put(VanillaHudElements.ARMOR_BAR, HEALTH_BAR);
+		builder.put(VanillaHudElements.MOUNT_HEALTH, ZERO);
+		builder.put(VanillaHudElements.FOOD_BAR, ZERO);
+		builder.put(VanillaHudElements.AIR_BAR, reduceToIntFunctions(MOUNT_HEALTH, FOOD_BAR, Integer::sum));
+		VANILLA_HEIGHT_PROVIDERS = builder.build();
+		LEFT_HEIGHT_PROVIDERS.put(VanillaHudElements.HEALTH_BAR, HEALTH_BAR);
+		LEFT_HEIGHT_PROVIDERS.put(VanillaHudElements.ARMOR_BAR, ARMOR_BAR);
+		RIGHT_HEIGHT_PROVIDERS.put(VanillaHudElements.MOUNT_HEALTH, MOUNT_HEALTH);
+		RIGHT_HEIGHT_PROVIDERS.put(VanillaHudElements.FOOD_BAR, FOOD_BAR);
+		RIGHT_HEIGHT_PROVIDERS.put(VanillaHudElements.AIR_BAR, AIR_BAR);
+	}
+
+	@Override
+	public void onInitializeClient() {
+		ClientLifecycleEvents.CLIENT_STARTED.register((MinecraftClient minecraft) -> {
+			HudStatusBarHeightRegistryImpl.init();
+		});
+	}
+
+	public static void addLeft(Identifier id, ToIntFunction<PlayerEntity> heightProvider) {
+		if (resolvedHeightProviders == null) {
+			LEFT_HEIGHT_PROVIDERS.put(id, heightProvider);
+		} else {
+			throw new IllegalStateException("Height provider registry already frozen!");
+		}
+	}
+
+	public static void addRight(Identifier id, ToIntFunction<PlayerEntity> heightProvider) {
+		if (resolvedHeightProviders == null) {
+			RIGHT_HEIGHT_PROVIDERS.put(id, heightProvider);
+		} else {
+			throw new IllegalStateException("Height provider registry already frozen!");
+		}
+	}
+
+	public static int getHeight(Identifier id) {
+		// returns zero for invalid calls, can adjust this to throw instead
+		PlayerEntity player = MinecraftClient.getInstance().inGameHud.getCameraPlayer();
+		if (player != null) {
```

Comment:
Prefer to use the [early return pattern](https://medium.com/swlh/return-early-pattern-3d18a41bba8), it makes the code much more readable. Also apples to the addRight/addLeft methods.